### ISSUE TITLE
SLEPc 3.6. requires to choose a linear solver in parallel run with ShiftInvert

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#1117]
although deal.II compiles fine with PETSc/SLEPc 3.6, i don't think there is interface in SLEPc solvers / spectral transformation to choose a linear solver for `TransformationShiftInvert`. 
That used to be a non-issue before 3.6, but now:
```
[0]PETSC ERROR: No support for this operation for this object type
[0]PETSC ERROR: You chose to solve linear systems with a factorization, but in parallel runs you need to select an external package; see the users guide for details
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.6.0, Jun, 09, 2015
```

So everything that use SLPEc in parallel will fail unless something like `-st_ksp_type gmres -st_pc_type gamg -st_ksp_rtol 1e-12` is provided from the command line.

**Comment by oneliefleft** [comment]
Thanks for reporting this Denis.

This looks like a parallel specific problem; right? Does your code work on
a single processor?
We'll have to look into modifying the SLEPcWrappers (SpectralTransformtion
part) to deal with this - possibly we can hijack SolverControl for this.
I am downloading the latest PETSc/SLEPc combo right now... and will have a
play.

Although this looks like a SLEPc specific problem, out of my own
curiousity, have you had any troubles using PETSc as a solver?

Best,
   Toby



2015-07-11 7:44 GMT+02:00 Denis Davydov <notifications@github.com>:

**Comment by davydden** [comment]
yes, it is indeed a MPI specific problem, more precisely when MPI matrices are used. To my understanding the reason is that by default SLEPc uses direct serial solver in ShiftAndInvert, and that is not good for parallel runs. Since `3.6` PETSc aborts such attempts.

I would say we need to provide an extra argument to either constructor of `TransformationShiftInvert` or via a function in `TransformationBase` to choose which PETSc solver will be used, which preconditioner and which tolerance. It would be nice to re-use classes from `PETScWrappers`, but i don't quite see yet how to do i because one would need to have access to `private` members of those classes to pass them to SLEPc. @bangerth  any advices? 

p.s. i didn't specifically test PETSc only, but `make test` (small test suite with 8 tests) of deal.II contains a test on PETSc, so I would say it works. There's also `step-17` which runs PETSc in parallel. Timo said tests pass on his computer, so it should be good.

**Comment by oneliefleft** [comment]
>
> yes, it is indeed a MPI specific problem, more precisely when MPI matrices
> are used.
>
 That is what it looks like to me too.

> I would say we need to provide an extra argument to either constructor of
> TransformationShiftInvert or via a function in TransformationBase to
> choose which PETSc solver will be used, which preconditioner and which
> tolerance. It would be nice to re-use classes from PETScWrappers, but i
> don't quite see yet how to do i because one would need to have access to
> private members of those classes to pass them to SLEPc. @bangerth
> <https://github.com/bangerth> any advices?
>
Yes and no. We should think about this a little. Yes, it would be nice to
reuse parts of the existing PETScWrappers though I do not think we need
access to private functions... (do we?) However, most of the auxiliary
functionality in SLEPcWrappers work by set_XXX () functions. Maybe we could
get away with a set_spectral_transformation_solver (something, something);
where something and something are a solver and SolverControl? I have to dig
in to the code a little deeper to remember how things work....

The small tests only check that basic parts of PETSc can be used; ie., that
the library links (uses CGSolver, for example). The entire test suite needs
to be run to know for sure. I will try to figure something out over the
week. I would not recommend changes like this before the 8.3 release
anyway, so we have time to come up with something that really works and
patch it in after.   :-)

Does that make sense?

Best,
   Toby

**Comment by davydden** [comment]
makes sense. 

i quickly looked through `SLEPc` documentation, so possible areas of linear solvers are
1) spectral transformations:
`-st_ksp_type preonly -st_pc_type lu -st_pc_factor_mat_solver_package mumps`
`-st_ksp_type preonly -st_pc_type lu -st_pc_factor_mat_solver_package superlu_dist`
the first one effectively mimics `PETScWrappers::SparseDirectMUMPS`.

`-st_ksp_type gmres -st_pc_type bjacobi -st_ksp_rtol 1e-9`

2) eigensolvers:
We need to allow this for some eigensolvers as well, i.e. Jacobi-Davidson:
`-eps_type jd -st_ksp_type gmres -st_pc_type jacobi -st_ksp_max_it 10`


To do it within the code, one needs to get KSP object first:
`STGetKSP(ST st,KSP *ksp);`
and then, supposedly (!), go on from there
`KSPGetPC(KSP ksp,PC *pc);`
`KSPSetType(KSP ksp,KSPType method);`
etc.

In order to reuse `PETScWrappers`, we would need to access private attribute `std_cxx11::shared_ptr< SolverData > 	solver_data`, which contains `KSP ksp` and  `PC pc`. Maybe declaring `SLEPcWrappers::SolverBase` to be a friend of `PETScWrappers::SolverBase` could solve this issue. There must be  a better way to do it from the object-oriented perspective...

.

**Comment by oneliefleft** [comment]
> 2) eigensolvers:
> We need to allow this for some eigensolvers as well, i.e. Jacobi-Davidson:
> -eps_type jd -st_ksp_type gmres -st_pc_type jacobi -st_ksp_max_it 10
>
> To do it within the code, one needs to get KSP object first:
> STGetKSP(ST st,KSP *ksp);
> and then, supposedly (!), go on from there
> KSPGetPC(KSP ksp,PC *pc);
> KSPSetType(KSP ksp,KSPType method);
> etc.
>

Now this reminds me of some experimental code I wrote a long time ago with
Eloy from SLEPc exactly for the Solver::GeneralizedDavidson (never got
round to putting it into the subversion though). Eigensolvers are easy. We
pass the dealii::PETScWrappers::SolverXXX and dealii::SolverData types in
as AdditionalData which is activated in the SLEPcWrappers::solve routines
by some additional functions, such as, set_ksp_type, etc. Essentially in
the way you wrote above with a few error checks for good measure.

Now I am wondering if we can simply add AdditionalData to all of the
SpectralTransformations. That would keep the interfaces uniform as well as
being completely backward-compatible.

How does that sound as a plan?

**Comment by davydden** [comment]
i don't quite understand yet why would you pass both `dealii::PETScWrappers::SolverXXX` and `dealii::SolverData`, as the latter is actually an attribute of `dealii::PETScWrappers::SolverBase`? If you could pull-out that chunk of code and set up a pull-request, it would be easier to discuss.

Some spectral transformation already have `AdditionalData`, so I guess then we just need to add two additional parameters there, i.e. `PETScWrappers::SolverBase` and `SolverControl`. And then do the magic inside, perhaps in a function implemented once in `TransformationBase`.
Whereas for eigensolvers which require linear solvers, `AdditionalData` would also be extended in the same way.

**Comment by tjhei** [comment]
Do you guys think you can work it out for the 8.3 release (in the next couple of days)? Note that we don't have any parallel slepc tests in our testsuite (and basically no other slepc tests for that matter):
```
$ grep -r -i slepc --include="*.cc" .
./slepc/00.cc:// just check initialising SLEPc can be done and that it initialises
./slepc/00.cc:#include <deal.II/lac/slepc_solver.h>
./slepc/00.cc:      logfile << "Initializing SLEPc (PETSc): "
./slepc/00.cc:      SlepcInitialize (&argc, &argv, 0, 0);
./slepc/00.cc:        logfile << "Finalizing SLEPc (PETSc): "
./slepc/00.cc:      SlepcFinalize ();
./quick_tests/step-slepc.cc:#include <deal.II/lac/slepc_solver.h>
./quick_tests/step-slepc.cc:// Test that deal.II is working with SLEPc by solving the Laplace's
./quick_tests/step-slepc.cc:  SLEPcWrappers::SolverArnoldi eigensolver (solver_control);
```
That is also something we should work on!

**Comment by oneliefleft** [comment]
>
> Do you guys think you can work it out for the 8.3 release (in the next
> couple of days)?
>
Sorry, I doubt it is possible. I expect in the next couple of weeks. At
least, I can not do much on this for a few days, though I am eager to make
it happen.

> Note that we don't have any parallel slepc tests in our testsuite (and
> basically no other slepc tests for that matter):
>
> That is also something we should work on!
>
Absolutely right! Not enough SLEPc tests in general. :-)

**Comment by tjhei** [comment]
> Sorry, I doubt it is possible. 

Okay, what do you suggest we do: recommend an older version, just ignore the issue altogether?

**Comment by davydden** [comment]
one does not need to recommend an older version. As long as a user choose linear solver from the command line (e.g. `-st_ksp_type gmres -st_pc_type bjacobi -st_ksp_rtol 1e-9`), it's all good. It just there is not mechanism yet to do it within the deal.II.

**Comment by davydden** [comment]
and this applies to MPI matrices only as SLEPc's default direct lu is good for serial matrices.

**Comment by oneliefleft** [comment]
one does not need to recommend an older version. As long as a user choose
> linear solver from the command line (e.g. -st_ksp_type gmres -st_pc_type
> bjacobi -st_ksp_rtol 1e-9), it's all good. It just there is not mechanism
> yet to do it within the deal.II.
>
Exactly right. Command-line options always work with PETSc and SLEPc. It is
just that we do not have a happy user-friendly interface to do it deal.II
style. We will make this work in due course. That, I can promise.

@timo My recommendation is to quietly ignore this issue on the next
release, though I would like to build up a better testsuite for SLEPc. The
discussion is here on the forum; grep and google are our friends. ;-) We
can very safely say that deal.II 8.3 has been tested with PETSc and SLEPc
3.5 and works out-of-the-box - I have no doubt that we can keep
backward-compatibility while fixing this issue. Does that sound ok with you?

Best,
   Toby

**Comment by tjhei** [comment]
> Does that sound ok with you?

Sure. ;-)

**Comment by yoczhang** [comment]
Hi, i want to know issue is solved？ i meet the same question, but in an other FEM programmes.

**Comment by davydden** [comment]
@yoczhang no, this issue is not resolved yet. I plan to look at it next month...

@tamiko i suppose https://github.com/dealii/experimental is a good place for collaborative efforts? I guess `master` branch should mirror `master` from `dealii/dealii`? I wonder if one can automate mirroring by something like `--set-upstream-to=dealii/master`?

**Comment by davydden** [comment]
@oneliefleft Toby, do you recall why you put all the initialisation and creation of SLEPc objects inside `SolverBase::solve` class? Looking at it now i would go for:

*  put EPSCreate() to constructors and in derived constructor do `EPSSetType()`.
*  shift `EPSSetOperators(eps,A,B)` and `EPSSolve(eps)` to the `Wrappers::solve()`
*  do `JacobiDavidson::set_initial_vectors()` and `KrylovSchur::set_initial_vector()` (not plural/single)
*  things like `EPSSetTarget` and `EPSSetWhichEigenpairs` would go to function of the base class where  they are called. Default values can be set in the base class constructor.

let me know if you have some comments, i will start looking at it these days as I want to have JD with starting eigenspace (multiple eigenvectors) and perhaps with some other features missing.

**Comment by oneliefleft** [comment]
Hi Denis,

Let's recall that I based the functionality of SLEPcWrappers on the
PETScWrappers - which may not have been the smartest decision - and I do
admit that the SLEPcWrappers could do with a little more love...   :-)

>
>    - put EPSCreate() to constructors and in derived constructor do
>    EPSSetType().
>    - shift EPSSetOperators(eps,A,B) and EPSSolve(eps) to the
>    Wrappers::solve()
>
>  Can you please elaborate on this? What is the advantage of this change? I
do see a certain logic to your suggestion, but maybe I am not quite
grasping it.

>
>    - do JacobiDavidson::set_initial_vectors() and
>    KrylovSchur::set_initial_vector() (not plural/single)
>
> KrylovSchur can only take one vector as an initial subspace and the
Davidson solvers can take many. It would be smarter to be generic and use
the plural since, if I remember correctly, if more than one eigenvector is
given to the KrylovSchur as an initial subspace only the first will be used
and the rest ignored (by SLEPc). If I do not remember correctly, it would
be easy to implement this behaviour in the wrappers. I would be happy to
have the plural the norm and deprecate the singular.

I am still in favour of having set_initial_vector somehow set in the
AdditionalData routine....

Best,
   Toby

**Comment by davydden** [comment]
>   Can you please elaborate on this? What is the advantage of this change? I
do see a certain logic to your suggestion, but maybe I am not quite
grasping it.

Well, i think it is a bit more transparent from the object-oriented point-of-view. So if we wrap PETSc objects in our classes, the creation of the object (`EPSCreate`) is done in the constructor of the class, followed by a specification inside a derived-class constructor (`EPSSetType`). And then things like `EPSSetOperators` can be done directly when they are called, usually in `solve(A,B,n_eigenvalues)`;  

The current implementation of `SLEPcWrappers` does everything in the long and winding ~~road~~ function `SolverBase::solve` which prevents any pre-initialisation of `EPS` object and it's re-usal, as far as i can tell. It also makes the whole class a bit obscure because when one calls it's methods like `set_target_eigenvalue`, the provided value is first cached somewhere else (i.e. shared-pointer object) to be later applied in that long `SolverBase::solve` function.

> KrylovSchur can only take one vector as an initial subspace and the Davidson solvers can take many.

sure, but at the end of the day those are different classes which may have extra special functions/functionalities. So I would actually intentionally do those specific "settings" in the derived classes. Maybe `Lanczos` or `power` or whatever else method can not use any info from be it multiple or single initial eigenvectors, then this "setting" function should not be available for it. 

@tamiko any comments from your side? I recall you had some ideas about (p)Arpack class. So maybe we should decide on interface all at once to make it consistent between different libraries.


Let me draft some changes in a separate PR this week and then we can continue from there...

**Comment by davydden** [comment]
let's move the discussion here https://github.com/dealii/dealii/pull/1804

**Comment by davydden** [comment]
@yoczhang the issue is being resolved in https://github.com/dealii/dealii/pull/1804 . So far it works for Jacobi preconditioner, i would appreciate help with testing.

**Comment by davydden** [comment]
the issue is fixed in the mentioned in the above PR.
